patches/minecraft/net/minecraft/world/entity/Entity.java.patch

```diff
@@ -371,6 +415,24 @@
 +      return this.m_6380_(pose, size);
 +   }
 +
++   protected FluidState touchingFluid = net.minecraft.world.level.material.Fluids.f_76191_.m_76145_();
++   protected Object2DoubleMap<FluidState> customFluidHeight = new it.unimi.dsi.fastutil.objects.Object2DoubleOpenHashMap<>();
++   public void setInFluid(FluidState touchingFluid) {
++       this.touchingFluid = touchingFluid;
++   }
++   public FluidState getTouchingFluid() {
++       return this.touchingFluid;
++   }
++   public boolean isInFluid() {
++       return this.getTouchingFluid() != net.minecraft.world.level.material.Fluids.f_76191_.m_76145_() && this.getFluidHeight(this.getTouchingFluid()) > 0.0D;
++   }
++   public void addFluidHeight(FluidState state, double fluidHeight) {
++       this.customFluidHeight.put(state, fluidHeight);
++   }
++   public double getFluidHeight(FluidState state) {
++       return this.customFluidHeight.getDouble(state);
++   }
```

Comment:
I feel like it may make sense to put a reference of each of these methods in `IForgeEntity` so that javadocs can be added saying when to use them, especially for `setInFluid` as it almost feels like something that shouldn't be external but has to be because of how things are hardcoded in entities so it would be good for modders to know when to call it.